File: 4- Elementry Search/Ternary Search/RecursiveTernary.py
class RecursiveTernary:
    @staticmethod
    def Search(low, high, key, Iterable):
        if high >= low:
            mid1 = low + (high - low) // 3
            mid2 = high - (high - low) // 3
            if Iterable[mid1] == key:
                return mid1
            if Iterable[mid2] == key:
                return mid2
            if key < Iterable[mid1]:
                return RecursiveTernary.Search(low, mid1 - 1, key, Iterable)
            elif key > Iterable[mid2]:
                return RecursiveTernary.Search(mid2 + 1, high, key, Iterable)
            else:
                return RecursiveTernary.Search(mid1 + 1, mid2 - 1, key, Iterable)
        return -1

File: 4- Elementry Search/Ternary Search/RecursiveTernary.cpp
#include <vector>
#include <cstdlib>
#include <ctime>
using namespace std;
class RecursiveTernary {
public:
    template<typename T>
    static int Search(vector<T> it, int low, int high, T key) {
        if (high >= low) {
            int mid1 = low + (high - low) / 3;
            int mid2 = high - (high - low) / 3;
            if (it[mid1] == key) {
                return mid1;
            }
            if (it[mid2] == key) {
                return mid2;
            }
            if (key < it[mid1]) {
                return RecursiveTernary::Search(it, low, mid1 - 1, key);
            } else if (key > it[mid2]) {
                return RecursiveTernary::Search(it, mid2 + 1, high, key);
            } else {
                return RecursiveTernary::Search(it, mid1 + 1, mid2 - 1, key);
            }
        }
        return -1;
    }
};